java/vector/src/main/java/org/apache/arrow/vector/complex/ListViewVector.java

```diff
@@ -0,0 +1,528 @@
+    return field;
+  }
+
+  @Override
+  public MinorType getMinorType() {
+    return MinorType.LIST;
+  }
+
+  @Override
+  public void clear() {
+    super.clear();
+    validityBuffer = releaseBuffer(validityBuffer);
+    lastSet = -1;
+  }
+
+  @Override
+  public void reset() {
+  }
+
+  @Override
+  public ArrowBuf[] getBuffers(boolean clear) {
+    return new ArrowBuf[0];
+  }
+
+  @Override
+  public List<?> getObject(int index) {
+    if (isSet(index) == 0) {
+      return null;
+    }
+    final List<Object> vals = new JsonStringArrayList<>();
+    final int start = offsetBuffer.getInt(index * OFFSET_WIDTH);
+    final int end = start + sizeBuffer.getInt((index) * SIZE_WIDTH);
+    final ValueVector vv = getDataVector();
+    for (int i = start; i < end; i++) {
+      vals.add(vv.getObject(i));
+    }
+
+    return vals;
+  }
+
+  @Override
+  public boolean isNull(int index) {
+    return (isSet(index) == 0);
+  }
+
+  @Override
+  public boolean isEmpty(int index) {
+    return false;
+  }
+
+  /**
+   * Same as {@link #isNull(int)}.
+   *
+   * @param index  position of element
+   * @return 1 if element at given index is not null, 0 otherwise
+   */
+  public int isSet(int index) {
+    final int byteIndex = index >> 3;
+    final byte b = validityBuffer.getByte(byteIndex);
+    final int bitIndex = index & 7;
+    return (b >> bitIndex) & 0x01;
+  }
+
+  @Override
+  public int getNullCount() {
+    return 0;
+  }
+
+  @Override
+  public int getValueCapacity() {
+    return 0;
+  }
+
+  private int getValidityAndSizeValueCapacity() {
+    final int offsetValueCapacity = Math.max(getOffsetBufferValueCapacity(), 0);
+    final int sizeValueCapacity = Math.max(getSizeBufferValueCapacity(), 0);
+    return Math.min(offsetValueCapacity, sizeValueCapacity);
+  }
+
+  private int getValidityBufferValueCapacity() {
+    return capAtMaxInt(validityBuffer.capacity() * 8);
+  }
+
+  @Override
+  public void setNull(int index) {
+
+  }
+
+  @Override
+  public int startNewValue(int index) {
+    while (index >= getValidityAndSizeValueCapacity()) {
+      reallocValidityAndSizeAndOffsetBuffers();
+    }
+
+    if (lastSet >= index) {
+      lastSet = index - 1;
+    }
+
+    if (index == 0) {
+      offsetBuffer.setInt(0, 0);
+    } else if (index > lastSet) {
+      for (int i = lastSet + 1; i <= index; i++) {
+        final int lastOffSet = offsetBuffer.getInt((i - 1L) * OFFSET_WIDTH);
+        final int lastSize = sizeBuffer.getInt((i - 1L) * SIZE_WIDTH);
+        final int newOffSet = lastOffSet + lastSize;
+        offsetBuffer.setInt(i * OFFSET_WIDTH, newOffSet);
+      }
+    } else {
+      final int lastOffset = offsetBuffer.getInt(lastSet * OFFSET_WIDTH);
+      final int lastSize = sizeBuffer.getInt(lastSet * SIZE_WIDTH);
+      final int newOffSet = lastOffset + lastSize;
+      offsetBuffer.setInt((lastSet + 1) * OFFSET_WIDTH, newOffSet);
+    }
```

Comment:
Shouldn't we always just set the offset to the end of the child vector?

Reply:
Last Set doesn't matter in this vector. A new list always begins (has its offset) at the end of the child vector.

Reply:
When user doesn't set the null explicitly but keep adding vectors to, for instance
0, 1, 3, 5, 6, 7 where 2, 4 are not set and they must be null (`setNull` is not used to do this in some test cases and use cases I saw with `ListVector`). I was merely using the `lastSet` to 
fill the offsets with accurate values. But, I see your point we can take look at the end of the child vector and set the offset for sure, but in this case can we? Either way we set `sizeBuffer` when we do `endList` maybe there we can fill the holes. 
What do you think?

Reply:
Something like the following could show case the use case and accuracy (please check if the accuracy of this)

```java
@Test
  public void testImplicitNullVectors() {
    try (ListViewVector listViewVector = ListViewVector.empty("sourceVector", allocator)) {
      UnionListViewWriter listViewWriter = listViewVector.getWriter();
      /* allocate memory */
      listViewWriter.allocate();

      final ArrowBuf offSetBuffer = listViewVector.getOffsetBuffer();
      final ArrowBuf sizeBuffer = listViewVector.getSizeBuffer();

      /* write the first list at index 0 */
      listViewWriter.setPosition(0);
      listViewWriter.startList();

      listViewWriter.bigInt().writeBigInt(12);
      listViewWriter.bigInt().writeBigInt(-7);
      listViewWriter.bigInt().writeBigInt(25);
      listViewWriter.endList();

      int offSet0 = offSetBuffer.getInt(0 * BaseRepeatedValueViewVector.OFFSET_WIDTH);
      int size0 = sizeBuffer.getInt(0 * BaseRepeatedValueViewVector.SIZE_WIDTH);
      int lastSet0 = listViewVector.getLastSet();

      // after the first list is written,
      // the initial offset must be 0,
      // the size must be 3 (as there are 3 elements in the array),
      // the lastSet must be 0 since, the first list is written at index 0.

      assertEquals(0, offSet0);
      assertEquals(3, size0);
      assertEquals(0, lastSet0);

      listViewWriter.setPosition(5);
      listViewWriter.startList();

      // writing the 6th list at index 5,
      // and the list items from index 1 through 4 are not populated.
      // but since there is a gap between the 0th and 5th list, in terms
      // of buffer allocation, the offset and size buffers must be updated
      // to reflect the implicit null vectors.

      for (int i = 1; i < 5; i++) {
        int offSet = offSetBuffer.getInt(i * BaseRepeatedValueViewVector.OFFSET_WIDTH);
        int size = sizeBuffer.getInt(i * BaseRepeatedValueViewVector.SIZE_WIDTH);
        int lastSet = listViewVector.getLastSet();
        // Since the list is not written, the offset and size must equal to child vector's size
        // i.e., 3, and size should be 0 as the list is not written.
        // And the last set value is the value currently being written, which is 5.
        assertEquals(3, offSet);
        assertEquals(0, size);
        assertEquals(5, lastSet);
      }

      listViewWriter.bigInt().writeBigInt(12);
      listViewWriter.bigInt().writeBigInt(25);
      listViewWriter.endList();

      int offSet5 = offSetBuffer.getInt(5 * BaseRepeatedValueViewVector.OFFSET_WIDTH);
      int size5 = sizeBuffer.getInt(5 * BaseRepeatedValueViewVector.SIZE_WIDTH);
      int lastSet5 = listViewVector.getLastSet();

      assertEquals(3, offSet5);
      assertEquals(2, size5);
      assertEquals(5, lastSet5);

      listViewWriter.setPosition(10);
      listViewWriter.startList();

      // writing the 11th list at index 10,
      // and the list items from index 6 through 10 are not populated.
      // but since there is a gap between the 5th and 11th list, in terms
      // of buffer allocation, the offset and size buffers must be updated
      // to reflect the implicit null vectors.
      for (int i = 6; i < 10; i++) {
        int offSet = offSetBuffer.getInt(i * BaseRepeatedValueViewVector.OFFSET_WIDTH);
        int size = sizeBuffer.getInt(i * BaseRepeatedValueViewVector.SIZE_WIDTH);
        int lastSet = listViewVector.getLastSet();
        // Since the list is not written, the offset and size must equal to child vector's size
        // i.e., 3, and size should be 0 as the list is not written.
        // And the last set value is the value currently being written, which is 10.
        assertEquals(5, offSet);
        assertEquals(0, size);
        assertEquals(10, lastSet);
      }

      listViewWriter.bigInt().writeBigInt(12);
      listViewWriter.endList();

      int offSet11 = offSetBuffer.getInt(10 * BaseRepeatedValueViewVector.OFFSET_WIDTH);
      int size11 = sizeBuffer.getInt(10 * BaseRepeatedValueViewVector.SIZE_WIDTH);
      int lastSet11 = listViewVector.getLastSet();

      assertEquals(5, offSet11);
      assertEquals(1, size11);
      assertEquals(10, lastSet11);

      listViewVector.setValueCount(11);
    }
  }
```

Reply:
@lidavidm the question is do we need a pointer to keep track of the last set index? AFAIU keeping it is neater and easy to understand, but is there a better alternative? I thought about the `validityBuffer` usage too, but it felt like not that neat.

Reply:
I suppose it's necessary if we have to support skipping indices, but I would only use it for filling in gaps and would not use it to determine the offset of the new item.

Reply:
Agreed! I will add more information in the comment section too.